samples/mpsl/timeslot/src/main.c

```diff
@@ -0,0 +1,229 @@
+ */
+
+#include <kernel.h>
+#include <console.h>
+#include <string.h>
+#include <misc/printk.h>
+#include <zephyr/types.h>
+#include <irq.h>
+#include <dk_buttons_and_leds.h>
+
+#include <mpsl_timeslot.h>
+#include <mpsl.h>
+
+#define RUN_STATUS_LED               DK_LED1
+#define TIMESLOT_STATUS_LED          DK_LED2
+
+#define RUN_STATUS_BLINK_INTERVAL_US (1000)
+#define TIMESLOT_REQUEST_DISTANCE_US (1000000)
+#define TIMESLOT_LENGTH_US           (200)
+
+#define MPSL_IRQ_LOW_PRIO            (4)
+#define STACKSIZE                    CONFIG_IDLE_STACK_SIZE
+#define THREAD_PRIORITY              K_LOWEST_APPLICATION_THREAD_PRIO
+
+volatile int timeslot_counter = 0;
+bool request_in_cb = true;
+
+/* Timeslot requests */
+mpsl_timeslot_request_t timeslot_request_earliest = {
+	.request_type = MPSL_TIMESLOT_REQ_TYPE_EARLIEST,
+	.params.earliest.hfclk = MPSL_TIMESLOT_HFCLK_CFG_NO_GUARANTEE,
+	.params.earliest.priority = MPSL_TIMESLOT_PRIORITY_NORMAL,
+	.params.earliest.length_us = TIMESLOT_LENGTH_US,
+	.params.earliest.timeout_us = 1000000
+};
+mpsl_timeslot_request_t timeslot_request_normal = {
+	.request_type = MPSL_TIMESLOT_REQ_TYPE_NORMAL,
+	.params.normal.hfclk = MPSL_TIMESLOT_HFCLK_CFG_NO_GUARANTEE,
+	.params.normal.priority = MPSL_TIMESLOT_PRIORITY_NORMAL,
+	.params.normal.distance_us = TIMESLOT_REQUEST_DISTANCE_US,
+	.params.normal.length_us = TIMESLOT_LENGTH_US
+};
+
+mpsl_timeslot_signal_return_param_t signal_callback_return_param;
+
+/* FIFO and FIFO data type used for printing the signal type from timeslot callback */
+K_FIFO_DEFINE(cb_fifo);
+struct cb_fifo_data_t {
+	void *fifo_reserved; /* 1st word reserved for use by fifo */
+	int cb_signal_type;
+};
+
+static void error(void)
+{
+	printk("ERROR!\n");
+	dk_set_leds_state(DK_ALL_LEDS_MSK, DK_NO_LEDS_MSK);
+
+	while (true) {
+		/* Spin for ever */
+		k_sleep(1000);
+	}
+}
+
+static mpsl_timeslot_signal_return_param_t *mpsl_timeslot_callback(uint32_t signal_type)
+{
+	mpsl_timeslot_signal_return_param_t *p_ret_val = NULL;
+
+	switch (signal_type) {
+	case MPSL_TIMESLOT_SIGNAL_START:
+		/* Toggle led on every timeslot START */
+		dk_set_led(TIMESLOT_STATUS_LED, (++timeslot_counter) % 2);
+
+		if (request_in_cb) {
+			/* Request new timeslot when callback returns */
+			signal_callback_return_param.params.request.p_next = &timeslot_request_normal;
+			signal_callback_return_param.callback_action = MPSL_TIMESLOT_SIGNAL_ACTION_REQUEST;
+		} else {
+			/* No callback action when callback returns, so timeslot will be ended */
+			signal_callback_return_param.params.request.p_next = NULL;
+			signal_callback_return_param.callback_action = MPSL_TIMESLOT_SIGNAL_ACTION_NONE;
+		}
+
+		p_ret_val = &signal_callback_return_param;
+		break;
+	case MPSL_TIMESLOT_SIGNAL_CANCELLED:
+		break;
+	case MPSL_TIMESLOT_SIGNAL_BLOCKED:
+		break;
+	case MPSL_TIMESLOT_SIGNAL_SESSION_CLOSED:
+		break;
+	case MPSL_TIMESLOT_SIGNAL_SESSION_IDLE:
+		break;
+	case MPSL_TIMESLOT_SIGNAL_INVALID_RETURN:
```

Comment:
Might be good to return error on all the signals that are not supposed to be received.
Add the rest of the signals to the switch case, timer0, radio, and etc.